clang/runtime/dpct-rt/include/debug/schema.hpp

```diff
@@ -0,0 +1,410 @@
+  }
+  hex_str = ss.str();
+  hex_str.erase(hex_str.rfind(','));
+}
+
+void copy_mem_to_device(void *dst, void *src, size_t size) {
+// To do: To enable clang++ compiler supported CUDA code.
+#ifdef __NVCC__
+  cudaMemcpy(dst, src, size, cudaMemcpyDeviceToHost);
+#else
+  dpct::get_default_queue().memcpy(dst, src, size).wait();
+#endif
+}
+
+inline bool is_device_point(void *p) {
+#ifdef __NVCC__
+  cudaPointerAttributes attr;
+  cudaPointerGetAttributes(&attr, p);
+  if (attr.type == cudaMemoryTypeDevice)
+    return true;
+  return false;
+#else
+  dpct::pointer_attributes attributes;
+  attributes.init(p);
+  if (attributes.get_device_pointer() != nullptr)
+    return true;
+  return false;
+#endif
+}
+
+inline void get_val_from_addr(std::string &value, std::shared_ptr<Schema> schema, void *addr,
+                                     size_t size) {
+  void *h_addr = addr;
+  if (is_device_point(addr)) {
+    h_addr = malloc(size);
+    copy_mem_to_device(h_addr, addr, size);
+  }
+  if (schema->is_basic_type()) {
+    value += schema->get_var_name() + ": ";
+    std::string hex_str = "";
+    get_data_as_hex(h_addr, size, hex_str);
+    value += hex_str + "\n";
+    if (is_device_point(addr))
+      free(h_addr);
+    return;
+  }
+  std::shared_ptr<Schema> type_schema = schema_map[schema->get_type_name()];
+
+  std::vector<std::shared_ptr<Schema>> type_members =
+      type_schema->get_type_member();
+  value += "The data of " + schema->get_var_name() + " is:\n";
+  for (auto member : type_members) {
+    value += member->get_var_name() + ": ";
+    std::string hex_str = "";
+    char *addr_with_offset = (char *)h_addr + member->get_offset();
+    if (member->is_basic_type()) {
+      get_data_as_hex((void *)addr_with_offset, member->get_val_size(),
+                      hex_str);
+    } else {
+      get_val_from_addr(hex_str, member, (void *)addr_with_offset, size);
+    }
+    value += hex_str + " \n";
+  }
+  if (is_device_point(addr))
+      free(h_addr);
+}
+
+inline void process_var(std::string &log) { log = ""; }
+
+template <class... Args>
+void process_var(std::string &log, const std::string &schema_str, long *value, size_t size,
+                        Args... args) {
+  std::shared_ptr<Schema> schema = parse_var_schema_str(schema_str);
+  if (schema == nullptr) {
+    error_exit("Cannot parse the variable schema, please double check.\n");
+  }
+  if (size == 0) {
+    size = schema->get_type_size();
+  }
+  set_size_of_schema(schema_str, size);
+  switch (schema->get_val_type()) {
+  case ValType::SCALAR:
+    get_val_from_addr(log, schema, (void *)value, size);
+    break;
+  case ValType::ARRAY:
+  case ValType::POINTER:
+    get_val_from_addr(log, schema, (void *)(*value), size);
+    break;
+  case ValType::POINTERTOPOINTER:
+    get_val_from_addr(log, schema, *(void **)(*value), size);
+    break;
+  };
+  std::string ret;
+  process_var(ret, args...);
+  log += ret;
+}
+
+static std::map<std::string, int> api_index;
```

Comment:
static std::map<std::string, int> api_index;
Defining variables in the header file may lead to redefinition, even if the variables are decorated with “static”